4422 global downtime notification

rianfowler opened the pull request:
## Description
- adds global downtime behavior to DowntimeNotification and ExternalServicesError components 

## Testing done
- verified downtime window behavior 
- _did not_ validate feature flag (couldn't get vets-api to work consistently locally) 

## Screenshots


## Todo
- [ ] add unit tests 
- [ ] add global service to remaining configs 

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Commit:
add downtime behavior

Commit:
add downtime behavior

Commit:
use feature toggle

Commit:
revert change to burials form config

rianfowler commented on src/platform/monitoring/DowntimeNotification/config/scheduledDowntimeWindow.js:
```diff
@@ -0,0 +1,6 @@
+
+export default {
+  downtimeStart: moment.tz('2020-01-07 00:01', 'America/New_York'),
+  downtimeEnd: moment.tz('2020-01-08 00:01', 'America/New_York'),
```
Configure downtime window here. Only supporting one global downtime window for now. 

This is a **fallback** and it's never used when the api is available.

bkjohnson replied:
So far this has made things difficult to test.  I'll do what I can to leave this pattern of a hardcoded downtime window while still allowing a test to change it.

ncksllvn replied:
I personally like the idea of making this a static JSON file

rianfowler replied:
@ncksllvn is there an easy way to represent time zones with a static json? i guess we could break this up into parts with the date, time, and timezone strings.

rianfowler commented on src/platform/monitoring/DowntimeNotification/actions/index.js:
```diff
@@ -50,18 +55,42 @@ export function dismissDowntimeWarning(appTitle) {
+          startTime: downtimeStart,
+          endTime: downtimeEnd,
+        })) ||
+      [];
```
these changes impact the DowntimeNotification component. Basically this generates the response data that the component would get from the api if _all of the services are down_ and adds it to a failure action.

rianfowler commented on src/platform/monitoring/DowntimeNotification/util/helpers.js:
```diff
@@ -26,6 +27,29 @@ export function getStatusForTimeframe(startTime, endTime) {
+  return [
+    {
+      attributes: {
+        externalService: 'global',
```
_most_ of our forms do not designate the services they depend on in the downtime config. decided to add a "global" value for those. Later on I'll point out how this gets piped through the config. 

this function generates a `/maintenance_windows` response data as if each service is down for the fallback window.

ncksllvn replied:
Any reason we have `global` as a floating string here instead of the constant?

rianfowler replied:
mostly just me getting it to work and quickly moving on. i'll update it.

rianfowler commented on src/platform/monitoring/DowntimeNotification/actions/index.js:
```diff
@@ -50,18 +55,42 @@ export function dismissDowntimeWarning(appTitle) {
     try {
       const response = await apiRequest('/maintenance_windows/');
-      data = response.data;
+      if (!response.ok) {
```
I wasn't able to verify that `apiRequest` returns this status. help is appreciated here.

bkjohnson replied:
It [doesn't look like](https://github.com/department-of-veterans-affairs/vets-website/blob/b7433ca8ac919f2616351bef8492e97095b7b4f6/src/platform/utilities/api/index.js#L92-L98) `apiRequest` returns the `ok` prop, only the json data.  However, if the response _isn't_ ok then it should reject the promise which we could handle in a `.catch()`

rianfowler replied:
alrighty- thanks for the info.

rianfowler commented on src/platform/monitoring/DowntimeNotification/actions/index.js:
```diff
@@ -50,18 +55,42 @@ export function dismissDowntimeWarning(appTitle) {
+      }
     } catch (err) {
-      // Probably in a test environment and the route isn't mocked.
-    } finally {
```
`finally` is always executed. I don't want to fire a success action if the `!response.ok`. I'm not sure how this will impact things. maybe @ncksllvn may know. 

I may not need to add a `localEnvironment` check. for now the feature toggle works. (i.e. if you don't run vets-api then the toggle value is false so by default you'll not see the downtime notification locally)

Commit:
remove exists check from reducer

rianfowler commented on src/platform/monitoring/DowntimeNotification/reducer/index.js:
```diff
@@ -16,6 +17,13 @@ const initialState = {
+        isReady: true,
+        isPending: false,
+        serviceMap: action.data && createServiceMap(action.data),
+      };
```
New reducer follows the same behavior as existing reducers.

rianfowler commented on src/applications/pre-need/config/form.jsx:
```diff
@@ -82,6 +83,9 @@ const formConfig = {
   trackingPrefix: 'preneed-',
   transformForSubmit: transform,
+  downtime: {
+    dependencies: [externalServices.global],
```
form configs that do not have services listed in a downtime config will use the catchall "global" service.

rianfowler commented on src/applications/pre-need/components/IntroductionPage.jsx:
```diff
@@ -27,6 +27,7 @@ class IntroductionPage extends React.Component {
           Eligibility for Burial in a VA National Cemetery).
         </p>
         <SaveInProgressIntro
+          downtime={this.props.route.formConfig.downtime}
```
in order to see the downtime notification, you have to pipe the downtime config into the `SaveInProgressIntro` component. It uses the DowntimeNotification component which requires a start and end time to work (pulled from the state).

rianfowler reviewed (commented):
process overview: 

1. call to `/maintenance_windows` fails 
2. check for global downtime window 
  a. if currently in global downtime, set downtime state for each service 
  b. if not in a global downtime, set no downtime state

rianfowler commented on src/platform/monitoring/external-services/actions.js:
```diff
@@ -7,9 +10,21 @@ export const LOADING_BACKEND_STATUSES = 'LOADING_BACKEND_STATUSES';
+      toggleValues(state).vaGlobalDowntimeNotification &&
+      moment().isAfter(downtimeStart) &&
+      moment().isBefore(downtimeEnd);
+    dispatch({
```
ExternalServicesError follows a slightly different pattern. it doesn't use windows- it deals with strictly a down or up state.

ncksllvn replied:
It looks like we could share some logic here with downtime notifications, but I don't know if we're trying to keep the implementations completely separate anyway.

rianfowler replied:
I added a ticket for reconciling this. The concerns overlap heavily. seems like the main difference is that one provides an arbitrary timeframe and one provides the current state (no start or end time). I don't think we need to have all of this plumbing for that sort of temporal difference.

Commit:
add forking to backend statuses reducer

rianfowler commented on src/platform/monitoring/external-services/reducer.js:
```diff
@@ -13,9 +13,21 @@ export default function externalServiceStatuses(state = INITIAL_STATE, action) {
+              {
+                service: 'Global',
+                serviceId: 'global',
+                status: 'maintenance',
```
if there is an active global downtime, then add the global status to backend statuses

ncksllvn replied:
Should we be using the `global` constant here instead of the string value?

rianfowler commented on src/platform/user/authentication/components/SignInModal.jsx:
```diff
@@ -90,7 +90,7 @@ class SignInModal extends React.Component {
         )}
         {this.downtimeBanner(
-          [EXTERNAL_SERVICES.mvi],
+          [EXTERNAL_SERVICES.mvi, EXTERNAL_SERVICES.global],
```
the signin modal provides specific content for each downtime type. I just reused the mvi copy because it's the most generic.

rianfowler reviewed (commented):
process overview 

1. call to `/backend_statuses` fails
2. check for global downtime window
a. if currently in global downtime, set downtime state for each service
b. if not in a global downtime, set no downtime state

bkjohnson committed:
Fix failing action unit test

The updated getScheduledDowntime return function now expects a state
argument to be there, and the apiRequest helper also expects the
resposne to contain a `url` otherwise it will break.

bkjohnson committed:
Fix getBackendStatuses unit tests

The function now checks a state variable and also makes a call to
dispatch to indicate if global downtime is active

bkjohnson committed:
Update tests for external services reducer

If the action specifies that the global downtime is active then the
status will change

bkjohnson committed:
Set state in action test to be empty objects

It doesn't matter what the state is as long as it exists.  Maybe it will
be worth using something like optional chaining instead of forcing
something to exist.

bkjohnson committed:
Also set state to be empty here like last commit

bkjohnson committed:
Add tests for createGlobalMaintenanceWindow helper

bkjohnson committed:
Refactor & add unit tests for getBackendStatuses

getBackendStatuses has been refactored to take a downtimeWindow as an
argument with the default being the hardcoded scheduled values.  Having
the argument allows us to pass in differen values for testing.

bkjohnson committed:
Properly mock apiRequest

bkjohnson committed:
Fix moment issues within tests

ncksllvn commented on src/platform/monitoring/DowntimeNotification/actions/index.js:
```diff
@@ -50,18 +55,42 @@ export function dismissDowntimeWarning(appTitle) {
+        dispatch({
+          type: ERROR_SCHEDULE_DOWNTIME,
+          data: globalDowntimeData,
+        });
```
Just some context - the error-handler was left empty before because if there was an error retrieving maintenance windows from the API, it would be too bold to make any assumptions off of that that would affect an app or the UI in general.

rianfowler replied:
I think we'll have some discussion about this soon because i think if the status api is down, then it's likely the whole site is down.

ncksllvn commented on src/platform/monitoring/DowntimeNotification/util/helpers.js:
```diff
@@ -26,6 +27,29 @@ export function getStatusForTimeframe(startTime, endTime) {
+        endTime,
+      },
+    })),
+  ];
```
Nit-picky, but I feel like this is something that should happen in state/reducer. As in - the action creator dispatches just the `global` downtime, and it's translated into state as downtime for all services. But conceptually, it would be cool if wasn't necessary to expand global downtime into each service (and so any consumer of downtime notifications always has `global` as a dependency) but I see why we're doing it this way.

rianfowler replied:
yeah good redux practice is to put everything you can into the reducer. expanding it in the reducer makes sense.

ncksllvn commented on src/applications/edu-benefits/5490/containers/IntroductionPage.jsx:
```diff
@@ -104,6 +105,7 @@ export class IntroductionPage extends React.Component {
         </div>
         <SaveInProgressIntro
           buttonOnly
+          downtime={this.props.route.formConfig.downtime}
```
It would be cool if this was a default prop instead, so that we don't have to worry about future or WIP forms adding this prop.

ncksllvn commented:
Looks good. Feel free to just ignore my comments since we're under time constraints.

bkjohnson committed:
Use .catch() in promise chain instead of res.ok

The apiRequest helper doesn't return the response status but it does
reject a promise if the response isn't ok.

I also think we may be able to remove the try/catch wrapper but I'm
leaving it in for now just because.

bkjohnson committed:
Try to mock fetch a different way

This is a mess.  I'm trying to add another test to verify that
ERROR_SCHEDULE_DOWNTIME is dispatched when a fetch response isn't ok, so
to do that I will need to be able to mock fetch so that it can succeed
in one test and "fail" in another.

bkjohnson committed:
Clean up a bit & reset the dispatcher

bkjohnson committed:
Add test to cover fetch response not being ok

bkjohnson committed:
Invoke downtime when response has error data

We are doing this because sometimes the api will respond with a 200 even
when there is a server error.

bkjohnson committed:
Adjust global downtime banner on sign-in modal

Disabling the buttons will come next

bkjohnson committed:
Disable the buttons on global downtime

This duplicates some code in ExternalServicesError - it may be worth
cleaning it up a bit

bkjohnson committed:
Fix some tests

Destructuring caused the function to be called apparently, and we are
only looking for it to be called once.

Additionally, sometimes the statuses prop doesn't exist so we need a
null check for that using optional chaining

Commit:
add downtime notification to gibct; update default downtime message for apps

rianfowler commented:
updated the generic DTN to use the copy from peggy and the alert box. added DTN to GIBCT. 

<img width="1041" alt="Screen Shot 2020-01-09 at 17 42 49" src="https://user-images.githubusercontent.com/4998130/72110971-7e998780-3307-11ea-9456-855180a059f8.png">
<img width="1041" alt="Screen Shot 2020-01-09 at 17 42 41" src="https://user-images.githubusercontent.com/4998130/72110972-7e998780-3307-11ea-8fde-23c3450160f6.png">

rianfowler commented on src/platform/monitoring/DowntimeNotification/actions/index.js:
```diff
@@ -50,18 +55,42 @@ export function dismissDowntimeWarning(appTitle) {
+    // default to empty array
+    const { downtimeStart, downtimeEnd } = scheduledDowntimeWindow;
+    const globalDowntimeData =
+      (toggleValues(state).vaGlobalDowntimeNotification &&
```
I think we can remove this now. I was going to use this to force it into the on state so we can test it in staging but i dont' think that's necessary now.

bkjohnson commented on src/platform/monitoring/external-services/ExternalServicesError.jsx:
```diff
@@ -20,12 +21,17 @@ class ExternalServicesError extends React.Component {
   };
 
   componentDidMount() {
+    const { onRender, statuses } = this.props;
```
I don't know where the `statuses` prop is coming from, but it works in the `render()` function

bkjohnson committed:
Add tests for the ExternalServiceError callback

The component now has a callback prop for when it renders.  This allows
parent components to know when it is rendered, which means that the
parent components can know when a service is down.  This is used to
disable login buttons

Commit:
Add global dependency to 0993-edu-benefits

Commit:
Add to 0994

bkjohnson committed:
Add test for downtimeWindow data

I feel like this could be done a bit better, but it verifies that when
downtime happens the global downtime data is the right shape

Commit:
Add to 996

Commit:
Add global dependency to 1990

Commit:
Add global dependency to 1990e

Commit:
Add global dependency to 1990n

Commit:
Add global dependency to 1995

bkjohnson committed:
Skip weird test for now

This fails on Jenkins and when `npm run test:unit` is run locally, but
if the specific file is run then it passes. Due to the time crunch I
don't think this is worth investigating at the moment

Commit:
Add global dependency to 5495

Commit:
Add downtime to 686's intro page

Commit:
Added global dependency to chapter31-vre

Commit:
Add global dependency to chapter36-vre

bkjohnson commented:
A big chunk of the missing test coverage is due to the `<SignInModal>` component which has no existing tests.  It had some modifications to it to add the global downtime notification as well as to disable the buttons during downtime.

Commit:
Add global dependency to feedback-tool

Commit:
Add global dependency to veteran-representative

Commit:
Add global dependency to vic-v2

Commit:
Merge branch '4422-global-downtime-notification' of github.com:department-of-veterans-affairs/vets-website into 4422-global-downtime-notification

cvalarida commented:
I added the `global` dependency to most forms.

## Known forms

-   [X] 0993-edu-benefits
    -   Introduction page doesn't render: /education/opt-out-information-sharing/opt-out-form-0993
-   [X] 0994-edu-benefits
    -   See `verifyRequiredPrefill` below
-   [X] 0996-higher-level-review
    -   See `verifyRequiredPrefill` below
-   [X] 1990-edu-benefits
-   [X] 1990e-edu-benefits
-   [X] 1990n-edu-benefits
-   [X] 1995-edu-benefits
-   [X] 526EZ-all-claims
    -   Already done: <https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/applications/disability-benefits/all-claims/config/form.js#L133>
    -   Doesn't use the global service, but it shows the banner as
        expected
-   [X] 526EZ-claims-increase
    -   Redirects to the 526EZ-all-claims app if:
        -   User isn't logged in OR
        -   User doesn't have a saved increase-only form
-   [X] 5490-edu-benefits
-   [X] 5495-edu-benefits
-   [X] 686-dependent-status
    -   Already had `evss` as a dependency
    -   Needed to add `downtime` as a prop to `AuthorizationComponent` as well as `SaveInProgress`
-   [X] burials
    -   Already had `icmhs` as a dependency
-   [X] chapter31-vre
-   [X] chapter36-vre
-   [X] hca
    -   Already had `es` as a dependency
-   [X] pensions
    -   Already had `icmhs` as a dependency
-   [X] pre-need
-   [ ] vaos
    -   Automatically redirects to login
-   [X] feedback-tool
-   [X] veteran-representative
-   [X] vic-v2



## Known issues
There were a few issues, however.
- VAOS redirects to sign in
-   Some forms have `verifyRequiredPrefill`
    -   If this is true, the `SaveInProgressIntro` will render the sign in
        bit before it gets a chance to check if it should render the
        DowntimNotification

Commit:
Merge branch '4422-global-downtime-notification' of github.com:department-of-veterans-affairs/vets-website into 4422-global-downtime-notification

Commit:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into 4422-global-downtime-notification

Commit:
revert downmessaging removal, remove feature flag

Commit:
add localhost check; fix linting issues

Commit:
disable unit test

Commit:
disable e2e test for downtime notification

Commit:
add correct downtime window

Commit:
fix backend statuses api failure action

Commit:
update time for test

Commit:
update time
